src/include/duckdb/common/lru_cache.hpp

```diff
@@ -0,0 +1,147 @@
+//===----------------------------------------------------------------------===//
+//                         DuckDB
+//
+// duckdb/common/lru_cache.hpp
+//
+//
+//===----------------------------------------------------------------------===//
+
+#pragma once
+
+#include <functional>
+
+#include "duckdb/common/common.hpp"
+#include "duckdb/common/list.hpp"
+#include "duckdb/common/mutex.hpp"
+#include "duckdb/common/shared_ptr.hpp"
+#include "duckdb/common/string.hpp"
+#include "duckdb/common/unordered_map.hpp"
+
+namespace duckdb {
+
+// A LRU cache implementation, whose value could be accessed in a shared manner with shared pointer.
+// Notice, it's not thread-safe.
+template <typename Key, typename Val, typename KeyHash = std::hash<Key>, typename KeyEqual = std::equal_to<Key>>
+class SharedLruCache {
+public:
+	using key_type = Key;
+	using mapped_type = shared_ptr<Val>;
+	using hasher = KeyHash;
+	using key_equal = KeyEqual;
+
+	// @param max_memory_p: Maximum total memory (in bytes) of entries. 0 means unlimited.
+	explicit SharedLruCache(idx_t max_memory_p) : max_memory(max_memory_p), current_memory(0) {
+	}
+
+	// Disable copy and move
+	SharedLruCache(const SharedLruCache &) = delete;
+	SharedLruCache &operator=(const SharedLruCache &) = delete;
+
+	~SharedLruCache() = default;
+
+	// Insert `value` with key `key` and explicit memory size. This will replace any previous entry with the same key.
+	void Put(Key key, shared_ptr<Val> value, idx_t memory_size) {
+		// Remove existing entry if present
+		auto existing_it = entry_map.find(key);
+		if (existing_it != entry_map.end()) {
+			current_memory -= existing_it->second.memory;
```

Comment:
Can we move the `current_memory -= ...;` into `DeleteImpl`? They are always called together, that way it's less error-prone